third_party/nvfuser/csrc/lower_loop_rotation.h

```diff
@@ -0,0 +1,141 @@
+#pragma once
+
+#include <executor_params.h>
+#include <ir_all_nodes.h>
+
+namespace nvfuser {
+
+// Note [Loop Rotation]
+// Loop Rotation is an optimization pass to improve instruction scheduling. For
+// a given loop, for example:
+//   for (int i = 0; i < n; i++) {
+//     line1(i);
+//     line2(i);
+//     line3(i);
+//     line4(i);
+//   }
+// If we rotate one line up, then we get
+//   if (0 < n) {
+//     line1(0);
+//   }
+//   for (int i = 0; i < n; i++) {
+//     line2(i);
+//     line3(i);
+//     line4(i);
+//     if (i + 1 < n) {
+//       line1(i + 1);
+//     }
+//   }
+// Similarly, if we rotate two lines up, then we get
+//   if (0 < n) {
+//     line1(0);
+//     line2(0);
+//   }
+//   for (int i = 0; i < n; i++) {
+//     line3(i);
+//     line4(i);
+//     if (i + 1 < n) {
+//       line1(i + 1);
+//       line2(i + 1);
+//     }
+//   }
+// In order to take advantage of this pass, the scheduler needs to specify which
+// loop to rotate and the consumers whose allocation and computation will be
+// rotated, and pass this information as compilation parameter. For example, if
+// I have a fusion that will create the following loop structure:
+//   for (int i = 0; i < id1.extent(); i++) {
+//     float T1[5];
+//     for (int j = 0; j < 5; j++) {
+//       if (i < T0.size[0]) {
+//         T1[j] = sin(T0[i, j]);
+//       }
+//     }
+//     float T2[5];
+//     for (int j = 0; j < 5; j++) {
+//       T2[j] = cos(T1[j]);
+//     }
+//     float T3[5];
+//     for (int j = 0; j < 5; j++) {
+//       T3[j] = exp(T2[j]);
+//     }
+//     for (int j = 0; j < 5; j++) {
+//       if (i < T4.size[0]) {
+//         T4[i, j] = log(T3[j]);
+//       }
+//     }
+//   }
+// Then the scheduler could make a compilation parameter {id1, {T1, T2}} to the
+// fusion, and this pass will transform the code as
+//   float T1[5];
+//   float T2[5];
+//   if (0 < id1.extent()) {
+//     for (int j = 0; j < 5; j++) {
+//       if (0 < T0.size[0]) {
+//         T1[j] = sin(T0[0, j]);
+//       }
+//     }
+//     for (int j = 0; j < 5; j++) {
+//       T2[j] = cos(T1[j]);
+//     }
+//   }
+//   for (int i = 0; i < id1.extent(); i++) {
+//     float T3[5];
+//     for (int j = 0; j < 5; j++) {
+//       T3[j] = exp(T2[j]);
+//     }
+//     for (int j = 0; j < 5; j++) {
+//       if (i < T4.size[0]) {
+//         T4[i, j] = log(T3[j]);
+//       }
+//     }
+//     if (i + 1 < id1.extent()) {
+//       for (int j = 0; j < 5; j++) {
+//         if (i + 1 < T0.size[0]) {
+//           T1[j] = sin(T0[i + 1, j]);
+//         }
+//       }
+//       for (int j = 0; j < 5; j++) {
+//         T2[j] = cos(T1[j]);
+//       }
+//     }
+//   }
+// Currently, because all our existing predicates should already cover
+// out-of-bound access, so we are omitting the predicates to get a
+// better-looking code:
+//   float T1[5];
+//   float T2[5];
+//   for (int j = 0; j < 5; j++) {
+//     if (0 < T0.size[0]) {
```

Comment:
I don't think so, we need support for zero sized tensor.

Reply:
So, shouldn't we only omit the outer predicate (`if (0 < id1.extent())`) only when each expression is fully predicated with no omission? That sounds like messy logic, though...

Reply:
Don't we already predicate all global and shared memory access? If it is access to register, I don't think this is a problem. We can do `T1[0] = sin(T0[0])` even if `T0` does not contain any value.

Reply:
The test in `test_gpu_loop_rotation.cpp` should ensure that they are predicated correctly

Reply:
And I do have a note in the test about something that expr simplifier should change:
https://github.com/csarofeen/pytorch/pull/2500/files#diff-1b518fe84a6b24ba92a3dd37a4164a3f15fc6dc8f885094c155426f36cf3e615R105-R110

Reply:
> Don't we already predicate all global and shared memory access? If it is access to register, I don't think this is a problem. We can do `T1[0] = sin(T0[0])` even if `T0` does not contain any value.

I don't think so. See https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/index_compute.cpp#L2657.

It only happens in a very trivial case with no split or merge.

Reply:
Yeah, I know that function. And my argument is, as long as we make sure that there is no out-of-bound memory access for global and shared memory, I don't really care if `if (0 < extent)` or `if (i + 1 < extent)` is checked or not. **So the code with and without this predicate will not be equivalent, but I don't care if it is equivalent or not.** Do we consider this claim safe? I started this PR with all these predicates, but I eventually removed them because I prefer to read the code that without it (this simplification is mostly for readability, not for perf, I didn't check perf of it).

Reply:
I agree with this:

> as long as we make sure that there is no out-of-bound memory access for global and shared memory, I don't really care if if (0 < extent) or if (i + 1 < extent) is checked or not.

What I'm asking is in some cases the predicate of an expr is omitted based on the assumption that the expr is located inside a loop. Since this transformation moves an expression out of a loop, this assumption may not be held, so my question is are we making sure there is no out-of-bound memory access for global and shared memory?

Reply:
Are you saying that `the predicate of an expr is omitted based on the assumption that the expr is located inside a loop`, and `canOmitStopPredicate` is the code that decide whether it should be omitted? I don't see any problem here: `canOmitStopPredicate` only omit the predicate if `stop_index + stop_offset < extent` is proved to be true, and if this is proved to be true, `0 < extent` or `i + 1 < extent` will automatically be satisfied?

Reply:
The omission is done based on the assumption that:

```
stop_index >= 0
stop_index < extent
```

This assumption is valid when `index_simple` is true (https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/index_compute.cpp#L2661). The analysis is that when the index has no defining expression, it must be true that the stop index should be just a loop index, so it is guaranteed to range from 0 to extent.

If this expr is placed outside the loop, we can no longer have the assumption. More specifically, when `extent` is 0, the loop should never execute, so the omission of the predicate of the expression shouldn't matter. However, when the expression is moved out of the loop, there should be some predicate not to execute the expression when `extent` is 0.

Am I missing something? I actually am not familiar with how extent-zero domains are handled, but it seems to me omitting the predicate is wrong in that case, granted it's super uncommon.

Reply:
I see what you mean. And I think what you said makes sense. Let me try to play with `canOmitStopPredicate` to see if we can avoid this issue.

> I actually am not familiar with how extent-zero domains are handled
I don't think extent-zero domains are handled. They are just treated no differently from other extents, so there might be bugs somewhere in our system.

Reply:
Do you think this commit fixes the problem? https://github.com/csarofeen/pytorch/pull/2500/commits/80d7763e141898771c64a5a9eea2d587dd509b62

Reply:
Yes, I think so. Thanks for the fix. Just left one minor comment.